pkg/column-mapping/column.go

```diff
@@ -0,0 +1,321 @@
+	info, err := m.queryColumnInfo(schema, table, columns)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if info.ignore == true {
+		return vals, nil
+	}
+
+	exp, ok := Exprs[info.rule.Expression]
+	if !ok {
+		return nil, errors.NotFoundf("column mapping expression %s", info.rule.Expression)
+	}
+
+	return exp(info, vals), nil
+}
+
+// HandleDDL handles ddl
+func (m *Mapping) HandleDDL(schema, table string, columns []string, statement string) (string, error) {
+	info, err := m.queryColumnInfo(schema, table, columns)
+	if err != nil {
+		return statement, errors.Trace(err)
+	}
+	if info.ignore == true {
+		return statement, nil
+	}
+
+	// only output erro now, wait fix it manually
+	return statement, errors.NotImplementedf("ddl %s @ column mapping rule %s/%s:%+v", statement, schema, table, info.rule)
+}
+
+func (m *Mapping) queryColumnInfo(schema, table string, columns []string) (*columnInfo, error) {
+	var info = &columnInfo{
+		ignore: true,
+	}
+
+	m.cache.RLock()
+	ci, ok := m.cache.infos[tableName(schema, table)]
+	m.cache.RUnlock()
+	if ok {
+		return ci, nil
+	}
+
+	rules := m.Match(schema, table)
+	if len(rules) == 0 {
+		m.cache.Lock()
+		m.cache.infos[tableName(schema, table)] = info
+		m.cache.Unlock()
+
+		return info, nil
+	}
+
+	var (
+		schemaRules []*Rule
+		tableRules  = make([]*Rule, 0, 1)
+	)
+	// classify rules into schema level rules and table level
+	// table level rules have highest priority
+	for i := range rules {
+		rule, ok := rules[i].(*Rule)
+		if !ok {
+			return nil, errors.NotValidf("column mapping rule %+v", rules[i])
+		}
+
+		if len(rule.PatternTable) == 0 {
+			schemaRules = append(schemaRules, rule)
+		} else {
+			tableRules = append(tableRules, rule)
+		}
+	}
+
+	// only support one expression for one table now, refine it later
+	var rule *Rule
+	if len(table) == 0 || len(tableRules) == 0 {
+		if len(schemaRules) > 1 {
+			return nil, errors.NotSupportedf("route %s/%s to rule set(%d)", schema, table, len(schemaRules))
+		}
+
+		if len(schemaRules) == 1 {
+			rule = schemaRules[0]
+		}
+	} else {
+		if len(tableRules) > 1 {
+			return nil, errors.NotSupportedf("route %s/%s to rule set(%d)", schema, table, len(tableRules))
+		}
+
+		if len(tableRules) == 1 {
+			rule = tableRules[0]
+		}
+	}
+	if rule == nil {
+		m.cache.Lock()
+		m.cache.infos[tableName(schema, table)] = info
+		m.cache.Unlock()
+
+		return info, nil
+	}
+
+	sourcePosition := findColumnPosition(columns, rule.SourceColumn)
+	targetPosition := findColumnPosition(columns, rule.TargetColumn)
+
+	sourcePosition, targetPosition, err := rule.checkColumnPosition(sourcePosition, targetPosition)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	info = &columnInfo{
+		sourcePosition: sourcePosition,
+		targetPosition: targetPosition,
+		rule:           rule,
+	}
+
+	m.cache.Lock()
+	m.cache.infos[tableName(schema, table)] = info
+	m.cache.Unlock()
+
+	return info, nil
+}
+
+func (m *Mapping) resetCache() {
```

Comment:
should we make it to be public? when executed DDL, column position may changed, then we may need to reset cache manually

Reply:
just call it in `HandleDDL`